### PR TITLE
Playerdamage: Adding Post EntityTakeDamage Hook

### PR DESCRIPTION
**Pull request by ZenBre4ker**
#657 
To be able to read out or manipulate the damage after TTT2 processed it, this simple hook is called after calculating the reduced damage to players due to karma.

### CODE REVIEW

**Comment by ZenBre4ker** [comment]
Can you use return hook.run() like this, so that hooks to "TTT2PostEntityTakeDamage" could also return a value to block the whole damage event? 
Or does this always lead to a block?

**Comment by TimGoll** [comment]
> Can you use return hook.run() like this, so that hooks to "TTT2PostEntityTakeDamage" could also return a value to block the whole damage event?
> Or does this always lead to a block?

This should work because it will pass `nil` as well. However You should test your PR before we review it.

**Comment by ZenBre4ker** [comment]
Finally back and the new Hook now is tested and allows to show damage decrease due to Karma. Returning this hook allows to cancel the damage event by returning a value.

Here's an example performed with a M98B from the M9k weapons, which deals 200 on the head and with 0 karma should deal 90% less, so only 20. See the effect ingame and the output in the server console, which states the real damage.
With old Hook:
Headshot with 0 Karma
grafik

With New Hook:
Headshot with 0 Karma
grafik
Headshot with 1000 Karma
grafik

File: gamemodes/terrortown/gamemode/server/sv_player.lua [hunk 1]
```diff
@@ -1477,3 +1481,15 @@ end
...
+-- Called by @{GM:EntityTakeDamage} after damage output was changed due to internal events. You can modify all parts of the damage info in this hook again.
+-- @param Entity ent The @{Entity} taking damage
+-- @param CTakeDamageInfo dmginfo Damage info
+-- @return boolean Return true to completely block the damage event
+-- @note e.g. no damage if a value after Karma calculation is too low, etc
+-- @hook
+-- @realm server
+-- @local
+function GM:TTT2PostEntityTakeDamage(ent, dmginfo)
+end
```
Comment (TimGoll) [inline gamemodes/terrortown/gamemode/server/sv_player.lua:1495]:
```suggestion
-- Called by @{GM:EntityTakeDamage} after damage output was changed due to internal events. You can modify all parts of the damage info in this hook again.
-- @param Entity ent The @{Entity} taking damage
-- @param DamageInfo dmgInfo The damage info
-- @return boolean Return true to completely block the damage event
-- @note There is no damage value if it is too low after the karma calculation.
-- @hook
-- @realm server
-- @local
function GM:TTT2PostEntityTakeDamage(ent, dmgInfo)

end
```

Comment (ZenBre4ker) [reply gamemodes/terrortown/gamemode/server/sv_player.lua]: The note was meant as an example what you could do with the post damage hook. But I removed that note instead.

**Comment by TimGoll** [comment]
Okay. It looks good to me. One thing though:
While it is ok that you can modify the damage in this hook, I don't think it is a good idea to mention or event emphasis on it. Doesn't modifying/canceling the damage in this hook defete the whole purpose why you added it?

**Comment by ZenBre4ker** [comment]
On one hand I also wasn't sure about that, because it would be nice if all changes would be made before TTT2 applies its changes and therefore only a copy of the data should be sent.
On the other hand an addon may want to change damage information after Karma has been calculated.

So I'm unsure what's better. Just not mentioning, that you can change the damage afterwards or just not passing the original reference and therefore making it read only.
But as I can't imagine any purpose of changing it afterwards instead of before addons, I should probably make that hook read-only.

What do you think?

**Comment by TimGoll** [comment]
In my opinion a read only hook is the way to go. Is there a scenario where you want to modify the damage after the karma? At least for my addons it doesn't matter. At least in my opinion armor and karma should be factored in AFTER all damage multipliers are calculated.
But that's just my opninion. What's your opinion @Alf21 @saibotk @Histalek ?

**Comment by ZenBre4ker** [comment]
Changed it, because I agree with you. Can't think of a good use, therefore making it safe to hook to this and read out the real damage.

File: gamemodes/terrortown/gamemode/server/sv_player.lua [hunk 2]
```diff
@@ -1210,7 +1210,8 @@ function GM:EntityTakeDamage(ent, dmginfo)
 	end
 
 	--Call a second Hook after Damage got processed to output the real Damage that has been dealt
-	return hook.Run("TTT2PostEntityTakeDamage", ent, dmginfo)
+	local tmpdmginfo =  table.Copy(dmginfo)
```
Comment (ZenBre4ker) [inline gamemodes/terrortown/gamemode/server/sv_player.lua:1213]: This is not possible, noticed it too late....

**Review by ZenBre4ker — COMMENTED** [review]
How do I copy dmgInfo though?

File: gamemodes/terrortown/gamemode/server/sv_player.lua (hunk 2 above)
Comment (TimGoll) [reply gamemodes/terrortown/gamemode/server/sv_player.lua:1213]: yes, it is probably userinfo, isn't it?

**Comment by ZenBre4ker** [comment]
Feature ist bereits in Garrys Mod enthalten:
https://wiki.facepunch.com/gmod/GM:PostEntityTakeDamage

**Pull request closed without merging**